### PR TITLE
Issue 54 - close attack vector

### PR DESCRIPTION
**Pull request by snd**
https://github.com/paritytech/parity-bridge/issues/54

- [x] figure out why tests pass locally and fail on travis
  - was stuck there for a bit
  - on travis we run the coverage tests. they insert additional statements into the contract code. that changes the gas cost. removed assertions that were sensitive to that
- [x] add storage var `estimatedGasCostOfWithdraw` to `HomeBridge`
- [x] modify `withdraw` as laid out in https://github.com/paritytech/parity-bridge/issues/54#issuecomment-351009494
- [x] find good estimate for `estimatedGasCostOfWithdraw`
- [x] extract functions that read values from message so they can be tested, reused and `withdraw` stays focused and is easier to read
- [x] implement `isValueInMessageLargeEnoughToCoverWithdrawCost`
- [x] make bridge processes (rust) call `isValueInMessageLargeEnoughToCoverWithdrawCost` and ignore messages for which it returns false (as laid out in https://github.com/paritytech/parity-bridge/issues/54#issuecomment-351009494)
  - this truly shuts down the attack
- [x] make transport tests run again
- [x] create transport test that doesn't sendTransaction when call to `isValueInMessageLargeEnoughToCoverWithdrawCost` returns false
- [x] add `estimatedGasCostOfWithdraw` to config of rust processes
- [x] keep/increase solidity test coverage
  - [x] test that values which don't cover costs are ignored
  - [x] test that cost is subtracted from value and paid to relaying authority
  - [x] test `isValueInMessageLargeEnoughToCoverWithdrawCost`
  - [x] test `getXFromMessage`, etc in isolation
- [x] documentation: readme section: `recipient pays for withdraw relay costs`
  - recommended gas cost estimate

### CODE REVIEW

**Comment by snd** [comment]
required modifications to the contracts are mostly complete. see todos above.

while there are still some todos it would already be great if someone could take a look at the code so i can address grumbles early.

i extracted some functions from the contract code. this results in more function calls and a slightly higher gas cost. i think the benefits, less duplication, more reuse, easier to test, easier to read, harder to break accidentally, outweigh the slight increase in gas.

@debris

File: bridge/src/bridge/withdraw_relay.rs [hunk 1]
```diff
@@ -46,7 +54,15 @@ fn signatures_payload(foreign: &foreign::ForeignBridge, signatures: u32, my_addr
...
-fn withdraw_relay_payload(home: &home::HomeBridge, signatures: Vec<Bytes>, message: Bytes) -> Bytes {
+fn message_value_sufficient_payload(home: &home::HomeBridge, message: &Bytes) -> Bytes {
+	assert_eq!(message.0.len(), 84, "ForeignBridge never accepts messages with len != 84 bytes; qed");
+	home
+		.functions()
+		.is_message_value_sufficient_to_cover_relay()
+		.input(message.0.clone()).into()
+}
+
+fn withdraw_relay_payload(home: &home::HomeBridge, signatures: &Vec<Bytes>, message: &Bytes) -> Bytes {
```
Comment (debris) [inline bridge/src/bridge/withdraw_relay.rs]: it's better to `&[Bytes]` instead of `&Vec<Bytes>`

File: contracts/bridge.sol [hunk 2]
```diff
@@ -99,63 +107,117 @@ contract HomeBridge {
...
+        assembly {
+            hash := mload(add(message, 84))
+        }
+        return hash;
+    }
+
+    /// to be called by authorities to check
+    /// whether they withdraw message should be relayed or whether it
+    /// is too low to cover the cost of calling withdraw and can be ignored
+    function isMessageValueSufficientToCoverRelay(bytes message) public view returns (bool) {
```
Comment (debris) [inline contracts/bridge.sol]: what is the difference between `const`,`pure` and `view`?

Comment (snd) [reply contracts/bridge.sol]: `constant` = `view` = does not modify state

`pure` = does not read or modify state

explained in more detail here:
https://solidity.readthedocs.io/en/develop/contracts.html#functions

Comment (snd) [reply contracts/bridge.sol]: `pure` = pure in the functional programming sense. same args -> same return every time

**Comment by snd** [comment]
> Why do we even create a Withdraw event on the ForeignChain if the message value is not sufficient to cover relay?

@debris the message value is in wei. on foreign chain we only know the estimated gas usage of the relay (call to withdraw) in gas. value is sufficient if `value > estimatedGasCostOfWithdraw * tx.gasPrice`. it depends on the gasPrice on home chain. we can't know that on foreign chain

see also:

> we could improve the user experience by requiring value to be sufficient in ForeignBridge.transfer. > that would make it much less likely for users to burn value on withdraws that would get ignored.
> ForeignBridge.transfer would need to know an estimate of the gasprice on home_chain. we could have an oracle for that.
> this is nice to have but not necessary.

https://github.com/paritytech/parity-bridge/issues/54#issuecomment-351009494

File: bridge/src/bridge/withdraw_relay.rs (hunk 1 above)
Comment (snd) [reply bridge/src/bridge/withdraw_relay.rs]: indeed. i'll change it

File: bridge/src/config.rs [hunk 3]
```diff
@@ -18,6 +18,7 @@ pub struct Config {
 	pub foreign: Node,
 	pub authorities: Authorities,
 	pub txs: Transactions,
+	pub estimated_gas_cost_of_withdraw: u32,
```
Comment (debris) [inline bridge/src/config.rs]: please also update `examples/config.json` and `README.md`

Comment (snd) [reply bridge/src/config.rs]: done

**Comment by snd** [comment]
@bjornwgnr this is functionally complete and well tested. i'll make another pass over it tomorrow and finish documentation. @debris has already looked at it so i expect only minor grumbles before merging.

**Comment by snd** [comment]
@debris added documentation. can i get a final review (especially proofreading of the documentation added to readme) please

File: contracts/bridge.sol [hunk 4]
```diff
@@ -99,63 +107,117 @@ contract HomeBridge {
...
-            // which is padding address to 32 bytes and reading recipient at offset 32.
-            // for more details see discussion in:
-            // https://github.com/paritytech/parity-bridge/issues/61
-            recipient := mload(add(message, 20))
-            value := mload(add(message, 52))
-            hash := mload(add(message, 84))
-        }
+        address recipient = getRecipientFromMessage(message);
+        uint value = getValueFromMessage(message);
+        bytes32 hash = getTransactionHashFromMessage(message);
```
Comment (debris) [inline contracts/bridge.sol]: These function calls are probably not inlined by solidity. Have you compared the gas cost of this function call on master branch vs this branch?

Comment (snd) [reply contracts/bridge.sol]: overall this PR increases the gas cost of `HomeBridge.withdraw` by `1634` (`2%`) from `71715` (master: https://travis-ci.org/paritytech/parity-bridge/jobs/318131779#L1221) to `73349` (snd-issue-54-attack-vector: https://travis-ci.org/paritytech/parity-bridge/jobs/326815388#L1279). this increase includes the function calls as well as the added functionality (calculation of cost and transfer to sender). imho this increase is  acceptable.

**Pull request merged**